### PR TITLE
WIP: BUG: sph_harm: fix behavior for large degrees

### LINKED ISSUE #7778: Overflow in lpmn and sph_harm for high degrees
I was trying to plot the EGM-96 geopotential model, which includes spherical harmonics to degree and order 360, like it is done here:

https://people.sc.fsu.edu/~lb13f/projects/space_environment/egm96.php

I first tried `special.lpmn`, but it overflows when the degree is high and therefore I cannot normalize it.

```python
In [2]: lpmn(360, 360, 0.0)[0][150, 150]
Out[2]: 3.7532741115719272e+306

In [3]: lpmn(360, 360, 0.0)[0][151, 151]
Out[3]: -inf
```

After some searching, I arrived to `sph_harm`, which seemed to be the right option, but apparently they overflow as well, for even lower orders:

```python
In [18]: sph_harm(85, 85, np.pi, 0.0)
Out[18]: -0j

In [19]: sph_harm(86, 86, np.pi, 0.0)
Out[19]: (nan+nan*j)
```

[The function `legendre_sphPlm` from GSL](https://www.gnu.org/software/gsl/manual/html_node/Associated-Legendre-Polynomials-and-Spherical-Harmonics.html), [also available in Octave](https://octave.sourceforge.io/gsl/function/gsl_sf_legendre_sphPlm.html), mostly does what I want, but it's GPL.

There are a variety of issues regarding the documentation or functionality of the Legendre functions in `scipy.special`, but still I am not sure if these overflows are implementation problems (features?) or whether I am missing something.

Relevant versions:

```python
In [21]: print(scipy.__version__, numpy.__version__, sys.version_info)
0.19.1 1.13.1 sys.version_info(major=3, minor=6, micro=2, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by ghost**
Closes #7778. 

@Juanlu001 can you confirm this?

### CODE REVIEW

File: scipy/special/sph_harm.pxd [hunk 1]
```diff
@@ -7,13 +7,20 @@ cdef extern from "c_misc/misc.h":
...
-    cdef double x, prefactor
+    cdef double x, prefactor, y
     cdef double complex val
     cdef int mp
+
+    if m == n and m > 65:
+        # Fall back to a sketchy implementation that returns
+        # a somewhat reasonable result, as the correct implementation fails
+        y = sqrt(1/M_PI)*0.5524*m**0.2428
+        return y*exp(m*phi*1j)*(sin(theta))**m
```
Comment (xoviat) [inline scipy/special/sph_harm.pxd]: What's the best way to cast to double complex?

**Comment by pv** [comment]
To be honest, I would prefer a PR with a complete fix implementing the appropriate asymptotic solution for any m,n, because that will likely be the least work in the long run.

Formulas used should be verifiable by someone else, for example with references or sufficient information on derivation given. Relative precision should be as close to machine precision as possible. Error larger than 1e-7 or so should generally result to nan values returned. New code paths should be verified by adding tests, for example against mpmath.

**Comment by Juanlu001** [comment]
@xoviat thanks for the effort. Haven't tried it yet for my particular case, but I agree with @pv that I would prefer something that I could verify from an independent source. "somewhat reasonable result" seems suspicious :)

@pv I guess these tests are what you're referring to: https://github.com/scipy/scipy/blob/v1.0.0/scipy/special/tests/test_mpmath.py

**Comment by xoviat** [comment]
>  New code paths should be verified by adding tests, for example against mpmath.

I agree with that part, but does mpmath support high-degree spherical harmonics? If so, then I may be able to numerically prove that the error of this approach is low enough.

**Comment by xoviat** [comment]
I guess it does.

**Comment by xoviat** [comment]
I guess `exp` doesn't support complex numbers.

**Comment by xoviat** [comment]
Either I'm doing something wrong, or the tests just passed. I still don't believe this implementation is that accurate, but @person142 or @pv can you check the test?

**Comment by xoviat** [comment]
Never mind; the test was marked slow.

**Comment by xoviat** [comment]
All things considered, the accuracy isn't bad:

```
E               Bad results (11 out of 2167) for the following points (in output 0):
E                                        100.0                           100.0                             0.0                             1.0 =>     (3.042238635984248e-08+0j) !=                            0.0  (rdiff                            0.0)
E                                        100.0                           100.0                           1e-30                             1.0 => (3.042238635984248e-08+3.042238635984248e-36j) !=                            0.0  (rdiff                            0.0)
E                                        100.0                           100.0           4.641588833612754e-21                             1.0 => (3.042238635984248e-08+1.4120820881969778e-26j) !=                            0.0  (rdiff                            0.0)
E                                        100.0                           100.0          2.1544346900318777e-11                             1.0 => (3.042238635984248e-08+6.554304452719726e-17j) !=                            0.0  (rdiff                            0.0)
E                                        100.0                           100.0                             0.1                             1.0 => (-2.552655824110764e-08-1.6550420423387113e-08j) !=         6.929398378656044e-101  (rdiff          4.390335884504723e+92)
E                                        100.0                           100.0                             1.0                             0.0 =>                             0j !=          3.029345023317297e-08  (rdiff                            1.0)
E                                        100.0                           100.0                             1.0                           1e-30 =>                             0j !=          3.029345023317297e-08  (rdiff                            1.0)
E                                        100.0                           100.0                             1.0           4.641588833612754e-21 =>                             0j !=          3.029345023317297e-08  (rdiff                            1.0)
E                                        100.0                           100.0                             1.0          2.1544346900318777e-11 =>                             0j !=          3.029345023317297e-08  (rdiff                            1.0)
E                                        100.0                           100.0                             1.0               6.283185307179586 =>                             0j !=          3.029345023317297e-08  (rdiff                            1.0)
E                                        100.0                           100.0              3.1415926535897927                             1.0 => (3.042238635984248e-08-1.6695511187284457e-21j) !=                            0.0  (rdiff                            0.0)
```

It looks like it's even under the given `1E-7` limit and I think can be slightly improved.

**Comment by pv** [comment]
It's 1e-7 relative error (the last rdiff column).

**Comment by xoviat** [comment]
Well we know when that happens actually.

**Comment by xoviat** [comment]
@pv How to return NAN?

**Comment by xoviat** [comment]
Wow. It actually passed the tests and agrees with mpmath. I can't believe it.

**Comment by Juanlu001** [comment]
Wow... Can't believe it either! I don't have the time to test this this week, but will do next one.

**Comment by xoviat** [comment]
So here's what I think @pv wants:

```
x = m*phi
a = (-1)**((m / 2)*(sgn(m) + 1))*(1j*sin(x) + cos(x)) / (abs(m)*gamma(m + 1)*2**(abs(m) + 1))
b = sqrt(((2*n + 1)*gamma(n + abs(m) + 1)) / (M_PI*(gamma(1 + n - abs(m)))))
c = (sin(theta))**(2*m / 2)
d = hyp2f1(abs(m) - n, n + abs(m) + 1, abs(m) + 1, (sin(phi/2))**2)

return a*b*c*d
```

Ref http://functions.wolfram.com/Polynomials/SphericalHarmonicY/26/01/02/0004/

**Comment by Juanlu001** [comment]
But won't that formula have the same problems? See https://github.com/scipy/scipy/issues/7778#issuecomment-323983599

**Comment by xoviat** [comment]
Possibly. That's what mpmath uses but it's arbitrary precision, and cython isn't. I just copied it down for the history.

**Comment by xoviat** [comment]
@Juanlu001 Ping.

**Comment by Juanlu001** [comment]
I will try to start again my original analysis and reproduce the plots in https://people.sc.fsu.edu/%7Elb13f/projects/space_environment/egm96.php, but that will have to wait for next month I fear.

**Comment by xoviat** [comment]
@Juanlu001 ping.

**Comment by xoviat** [comment]
I'm closing this. To be honest, I do not have the time to derive the asymptotic solution, and I don't see this moving forward.

**Comment by Juanlu001** [comment]
I apology for the delay in trying this, but I just haven't found the time to do it yet.

**Pull request closed without merging**